src/Interpreters/Session.cpp

```diff
@@ -294,6 +295,9 @@ AuthenticationType Session::getAuthenticationTypeOrLogInFailure(const String & u
 
 void Session::authenticate(const String & user_name, const String & password, const Poco::Net::SocketAddress & address)
 {
+    //Users with AuthType no_password or plaintext_password are not allowed to establish session if allow_plaintext_and_no_password is unset in server configuration.
+    if (!global_context->getAccessControl().getAuthTypeSetting() && (getAuthenticationType(user_name) == AuthenticationType::PLAINTEXT_PASSWORD || getAuthenticationType(user_name) == AuthenticationType::NO_PASSWORD))
+        throw Exception(ErrorCodes::AUTHENTICATION_DENIED , "Authentication denied for user configured with AuthenticationType::PLAINTEXT_PASSWORD and AuthenticationType::NO_PASSWORD. Please check with Clickhouse admin to allow allow_plaintext_and_no_plaintext_password");
```

Comment:
Let's move this logic to `IAccessStorage::authenticateImpl()`.
`getAuthenticationType()` is a bad function and should be avoided (it only exists because `MySQLHandler` hasn't been refactored yet).

Reply:
We can keep this check in are credentialsvalid() but we don't have global_context  access in IaccessStorage or AccessControl to read the config values.

Reply:
`AccessControl::authenticate()` calls `IAccessStorage::authenticate()` and can pass those two booleans to it.